src/include/cart/api.h

```diff
@@ -104,6 +104,36 @@ crt_init(crt_group_id_t grpid, uint32_t flags)
 int
 crt_context_create(crt_context_t *crt_ctx);
 
+
+/**
+ * Returns number of interfaces passed to CaRT at initialization time.
+ *
+ * Interfaces are passed via either D_INTERFACE environment variable or
+ * through crt_init_options_t::cio_interface options to crt_init_opt() .
+ *
+ * \return                     Number of interfaces CaRT is initialized with.
+ */
+uint32_t
+crt_num_ifaces_get(void);
+
+
+/**
+ * Create CRT transport context on an interface specified by the index.
+ * Index must be less than total number of interfaces returned by
+ * crt_num_ifaces_get().
+ *
+ * Must be destroyed by crt_context_destroy() before calling crt_finalize().
+ *
+ * Note: This is a client-side only API.
+ *
+ * \param[in]  iface_index     index of the interface.
+ * \param[out] crt_ctx         created CRT transport context
+ *
+ * \return                     DER_SUCCESS on success, negative value if error
+ */
+int
+crt_context_create_on_iface_idx(uint32_t iface_index, crt_context_t *crt_ctx);
```

Comment:
how can i get information on an interface given an index?
the way i would use this is that i would init cart with all possible interfaces, then i want to create a context with an interface where this context will be running on (the numa binding).

Reply:
The idea here is that as an app you wouldn't want to hardcode any interface names, so instead when you initialize CaRT with a list of interfaces you can then refer to them in a generic manner by the index in the order in which they were listed. 

The way I imagine daos layer would use new apis is something like:
`num_ifaces = crt_num_ifaces_get();
for (i = 0; i < num_ifaces; i++) {
    crt_context_create_on_iface_idx(i, &ctx);
    // Create event queue with this context
}
`

If it helps I can also add lookup functions to return an interface name by the index and vice versa as we already have this information stored internally. Beyond the name of the interface (as passed at init) we don't have any other information about it today in CaRT.

Reply:
no i was not talking about hardcoding of course.
in the end, we need to make some correlation between the iface index and the CPU of the core to select the proper one. i guess we can also keep that mapping cached on daos_init when talking to the agent first.
anyway might be a good idea to discuss this in more detail with Kris and sketch a use case out to just verify.

Reply:
Since daos_agent is already going to provide a client with the interface to be used (I assume it is based on the core on which app was running?), perhaps we can have an iface_name2indx API to do the conversion, and then subsequently use crt_context_create_on_iface_idx() to allocate the required one.  Do you think this will work?

Reply:
the gist of the problem today is that the agent provided one thing during init, but with multi threaded apps, the thread might be created on a different CPU/numa node, and so it needs a different one than what we selected on init.
So my thought of what we need is that on daos_init() we tell cart to use all interfaces (i guess the agent can provide those?), but it would be nice to also provide the ability to specify a default one to maintain original behavior like today with context_create without an iface idx.

Then on every context creation (with this new API), we need to specify the index of the iface/nic of the calling thread cpu affinity. this last part sounds like we need to ping the agent again for that info (agent would provide a string of the iface name I assume ?). but then we need to do the translation to that index.
so I see two potential issues:
1) pinging the agent on every context create - might be expensive.
2) the translation from the iface from the agent to the crt idx.
so for 2 if the iface_name2idx() is consistent / deterministic, then it should be good for that one.

Reply:
Perhaps then we can provide APIs such as crt_interface_select(char *) / crt_interface_select_by_idx(int) which would modify the default interface on which crt_context_create() will allocate the context?

name2idx is already just strcmps but with this then can be done only once if app needs to create multiple contexts on the same interface.

for 1. -- is the mapping of core to nic a fixed thing? I was assuming daos_agent does some sort of balancing as well to 'spread the load' across the nics.

If the mapping is fixed then perhaps we can query it from the agent once and do the translation ourselves (whether in daos or cart layer).
@kjacque Any thoughts on the agent interaction here?

Reply:
> for 1. -- is the mapping of core to nic a fixed thing? I was assuming daos_agent does some sort of balancing as well to 'spread the load' across the nics.

not always. on aurora for example there are 4 nics per CPU on the client side and the agent round robins in that case.